[Fahim Tajwar] iP

fahim-tazz opened the pull request:
# DukePro
> \"Your mind is for having ideas, not storing them.\" - [David Allen](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind off remembering tasks. It's  
- Text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

And it's **free**!

All you need to do is:
1. Download Duke from [here](https://github.com/fahim-tazz/ip/releases/tag/A-JAR).
2. Go to Terminal and enter the following:
```sh
cd ~/Downloads/         #or wherever your .jar file is.
java -jar duke.java
# Enjoy!
```
> Double-clicking the .jar file can corrupt it on MacOS, since MacOS flags it as unsafe.

3. Add your tasks with commands : `todo`, `deadline` and`event`.
4. Mark tasks as done or undone with `mark` and `unmark`.
5. Delete unwanted tasks with `delete`.
6. Can't find the task you're looking for? Just use `find`!
7. And let Duke handle the rest! :bowtie:

Features:
- [x] Managing todo-list
- [x] Managing deadlines
- [x] Tracking future events
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Commit:
Started with chatbot that imitates user commands, and exits on 'bye'.

Commit:
Added 'add' and 'list' functionality. Converted if-else-if tree to nested switch-case.

Commit:
Added Task class. Changes to Duke.java to store and print Task objects.

Commit:
Added 'mark' and 'unmark' functionality.

Commit:
Added line of text when calling 'list' command.

Commit:
Added ToDo, Event and Deadline tasks as subclasses of Task.

Commit:
Performed automated Text UI testing.

Commit:
Added exception handling for empty args, invalid command, bad indexes, badly formatted deadlines, badly formatted events.

Commit:
Added a DukeException class, as a parent class of all Duke-specific exceptions to come.

Commit:
Added delete functionality with exception handling.

Commit:
Style changes, add prevTask file detection.

Commit:
Add data.txt file reading functionality.

Commit:
Remove save/load functionality from Duke.java (transfer to FileHandler.java on next commit).

Commit:
Add FileHandler.java for save/load functionality.

ajjajjajjajj reviewed (commented):
Good work here, some suggestions for your code:
- Wildcard imports shouldn't be used, this might've resulted from the default settings in Intellij that auto converts imports into wildcard imports (try disabling it)
- You might want to implement an `symbol` attribute to each type of task, rather than hardcoding it into the `toString()` method. This would also help for the storage requirement, if you want to use the task symbols in the storage format :)
- A `reply(String reply)` method in your Duke class can solve the tediousness of having to print the divider line for each output by the chatbot
- Each case for a switch statement should be on the same indent line as the `switch():` statement itself, there's also a setting in Intellij that disables that

Nice job though, I found your code easy to read and understand :)

Commit:
Store Deadline dates as java.time.LocalDate.
Add error message for badly formatted dates.

Commit:
Merge branch 'branch-Level-7'

Commit:
Extract all command line handling from Duke.java into UI.java.

Commit:
Rename FileHandler.java to Storage.java

Commit:
Fix a bug where entering an invalid date created a Deadline with "by" value of NULL.

Commit:
Created separate welcome() method for welcome message.

jedidiahC commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,122 @@
+    ArrayList<Task> tasks = new ArrayList<>();
+    Scanner sc = new Scanner(System.in);
+
+    while(sc.hasNext()) {
```
I noticed several Java reserved words such as `while` in this case not being followed by a white space. e.g.
```suggestion
    while (sc.hasNext()) {
```
Perhaps you could consider using IntelliJ's code cleanup feature to add these white spaces automatically.

arkarsg replied:
Be careful of using `sc.hasNext()` vs `sc.hasNextLine()`

jedidiahC commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,122 @@
+    while(sc.hasNext()) {
+      String instr = sc.nextLine();
+      switch(instr) {                                 //Single-word instructions
+        case "list":  //List down
```
Should indentation for `case` clauses be removed?
```suggestion
    case "list":
```

jedidiahC commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,122 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+  public static void main(String[] args) {
```
Should basic indentation be changed from 2 spaces to 4 spaces to align with the Java coding standard?

```suggestion
    public static void main(String[] args) {

jedidiahC reviewed (commented):
Overall, the code seems okay. However, you may want to consider some of Java's coding standards in regards to layout and comments.

arkarsg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,122 @@
+            System.out.println(divider + "OOPS!!! The description of a " + action + " cannot be empty.\n" + divider);
+          }
+
+          switch(action) {                            //For instructions with argument(s).
```
Would recommend against nesting `switch-case` block for code quality

arkarsg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,122 @@
+          }
+      }
     }
+  }
```
A lot of nested blocks. Consider abstracting away with methods within the main class. For example you can have:
```java
// in main
switch(...) {
    case "todo":
        addToDo(tokens);
}

// add implementation
public void addToDo(String[] tokens) {
// your implementation
}
```

arkarsg reviewed (commented):
Good start so far. Consider abstracting away with more functions as the code expands with future levels.

Commit:
Redesign to maintain OO principles.
Add separate UI, Parser, TaskList classes.
Add NoArgumentException to handle empty argument cases.

Commit:
Group all classes into packages.

Commit:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

Commit:
Add Gradle support.

Commit:
Add JUnit testcases for TaskList.java.

Commit:
Modify TaskList.java to create new tasks.txt in the beginning if not already present.

Commit:
Make all backend classes package-private.

Commit:
Add Javadoc comments to duke.exceptions and duke.tasks package classes.

Commit:
Modify TaskListTest to clear tasks.txt before testing.

Commit:
Modify classes in duke.backend to conform to our Coding Standards.

Commit:
Make Event start and end times private.
Add getter method for Event start and end times.

Commit:
Modify duke.tasks classes to follow Coding Standard.

Commit:
Modify duke.exceptions classes to follow Coding Standard.

Commit:
Add Javadoc comments to duke.exceptions and duke.tasks package classes.

Commit:
Modify TaskListTest to clear tasks.txt before testing.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-codingStandard'. Resolved coding standards.

# Conflicts:
#	src/main/java/duke/exceptions/DukeException.java
#	src/main/java/duke/exceptions/NoArgumentException.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java
#	src/main/java/duke/tasks/Task.java
#	src/main/java/duke/tasks/Todo.java

Commit:
Fully Resolved merge conflicts between A-JavaDoc, A-codingStandard and master branches.

Commit:
Resolve some remaining merge conflicts in duke.exceptions.

Commit:
Add FIND keyword functionality.

Commit:
Add comments for UI:find() method.

Commit:
Set up Checkstyle for the project.

Commit:
Fix Checkstyle violations.

Commit:
Create GUI

Commit:
Add duke.commands package, containing all different types of commands.
Remove Ui.java.
Restructure the passage of input and output out of Duke object.

Commit:
Add functionality: close Duke on sending "bye" command.

Commit:
Fix CheckStyle Violations.

Commit:
Use assertions to ensure correct indexes are passed during Delete and Mark/Unmark.

fahim-tazz committed:
Merge pull request #2 from fahim-tazz/branch-A-codingStandard

Branch-A-CodingStandard

fahim-tazz committed:
Merge pull request #3 from fahim-tazz/branch-Level-9

Branch level 9

fahim-tazz committed:
Merge pull request #4 from fahim-tazz/branch-A-CheckStyle

Branch a check style

fahim-tazz committed:
Merge pull request #5 from fahim-tazz/branch-Level-10

Branch level 10

fahim-tazz committed:
Merge pull request #6 from fahim-tazz/branch-A-Assertions

Branch a assertions

Commit:
Add functionality: Detect tasks with duplicate names.

Commit:
Delete dead class Ui.java.
P.S. Checkstyle violations fixed in previous commits.

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Rename Duke to Jarvis.

Commit:
Add welcome message when GUI is first opened.

Commit:
'list' now says 'no items in list' when called on an empty list.

Commit:
Change User and Chatbot profile pictures.

Commit:
Rename Jarvis to Chad.

Commit:
Tweak Chad's responses.

Commit:
Change GUI background color.
Long lists are no longer cut short with "...".

Commit:
Fix infinite recursion on 'bye' command.

Commit:
Add JUnit tests for Parser.

Commit:
Remove redundant Sout debug messages from Storage.java.

Commit:
Set Gradle to build without checkstyle.

Commit:
Fix NullPointerException in all Make commands.

Commit:
Add all exception handling to Parser.java.

Commit:
Merge tag 'A-MoreErrorHandling'

Commit:
Change window label to "Chad - Your Personal DIgital Assistant".

Commit:
Add UI screenshot to docs folder.

Commit:
Add save functionality when window is closed by [X] button.

fahim-tazz committed:
Update README.md to include full User Guide.

Commit:
Reduce app closing delay after sending 'bye' command.

Commit:
Fix issue with not creating file. [Only works for command line execution].

Commit:
Update docs/README.md quick start to show how to launch through command line.

Commit:
Add user guide as README to root directory.

Commit:
Add Error Handling for invalid arguments for delete command.

fahim-tazz committed:
Fix hyperlink formatting in root/README.md

fahim-tazz committed:
Update README.md on root folder.

fahim-tazz committed:
Update README.md

Commit:
Cleanup code.
Remove redundant exception class.
Remove dead code.
